contract/models/abstract_contract_line.py

```diff
@@ -47,46 +48,32 @@ class ContractAbstractContractLine(models.AbstractModel):
         help="Sequence of the contract line when displaying contracts",
     )
     recurring_rule_type = fields.Selection(
-        [
-            ("daily", "Day(s)"),
-            ("weekly", "Week(s)"),
-            ("monthly", "Month(s)"),
-            ("monthlylastday", "Month(s) last day"),
-            ("quarterly", "Quarter(s)"),
-            ("semesterly", "Semester(s)"),
-            ("yearly", "Year(s)"),
-        ],
-        default="monthly",
-        string="Recurrence",
-        help="Specify Interval for automatic invoice generation.",
+        compute="_compute_recurring_rule_type",
+        store=True,
+        readonly=False,
         required=True,
+        copy=True,
     )
     recurring_invoicing_type = fields.Selection(
-        [("pre-paid", "Pre-paid"), ("post-paid", "Post-paid")],
-        default="pre-paid",
-        string="Invoicing type",
-        help=(
-            "Specify if the invoice must be generated at the beginning "
-            "(pre-paid) or end (post-paid) of the period."
-        ),
+        compute="_compute_recurring_invoicing_type",
+        store=True,
+        readonly=False,
         required=True,
-    )
-    recurring_invoicing_offset = fields.Integer(
-        compute="_compute_recurring_invoicing_offset",
-        string="Invoicing offset",
-        help=(
-            "Number of days to offset the invoice from the period end "
-            "date (in post-paid mode) or start date (in pre-paid mode)."
-        ),
+        copy=True,
     )
     recurring_interval = fields.Integer(
-        default=1,
-        string="Invoice Every",
-        help="Invoice every (Days/Week/Month/Year)",
+        compute="_compute_recurring_interval",
+        store=True,
+        readonly=False,
         required=True,
+        copy=True,
+    )
+    date_start = fields.Date(
+        compute="_compute_date_start", store=True, readonly=False, copy=True,
+    )
+    recurring_next_date = fields.Date(
+        compute="_compute_recurring_next_date", store=True, readonly=False, copy=True,
     )
```

Comment:
If fields are computed and `readonly=False`, shouldn't they have an inverse method? :thinking:

Reply:
No, this is the new API feature. What we have been calling "computed writable fields". They are like normal fields, but react to several depends and recompute its value when that dependencies are changed.